src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
+node's sequence number to 0 (a SHAMap never has a sequence number of 0). This
+is done for every node in the trie when `SHAMap::walkSubTree` is executed.
 
-*This description is obsolete and needs to be rewritten.*
+Note that other objects in rippled also have sequence numbers (e.g. ledgers).
+The SHAMap and node sequence numbers should not be confused with these other
+sequence numbers (no relation).
 
-SHAMaps can be thread safe, depending on how they are used.  The SHAMap
-uses a SyncUnorderedMap for its storage.  The SyncUnorderedMap has three
-thread-safe methods:
-
- * size(),
- * canonicalize(), and
- * retrieve()
+## SHAMap Creation ##
 
-As long as the SHAMap uses only those three interfaces to its storage
-(the mTNByID variable [which stands for Tree Node by ID]) the SHAMap is
-thread safe.
+A SHAMap is usually not created from vacuum.  Once an initial SHAMap is
+constructed, later SHAMaps are usually created by calling snapShot(bool
+isMutable) on the original SHAMap().  The returned SHAMap has the expected
+characteristics (mutable or immutable) based on the passed in flag.
 
+It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
+snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might be
+modified must be copied before they are placed in the mutable map.
+
+A new SHAMap is created with each new ledger round.  Transactions not executed
+in the previous ledger populate the SHAMap for the new ledger.
+
+## Storing SHAMap data in the database ##
+
+When consensus is reached, the ledger is closed.  As part of this process, the
+SHAMap is stored to the database by calling `SHAMap::flushDirty` with parameters
+which specify which nodes are to be written.
+
+Both `unshare()` and `flushDirty` walk the SHAMap by calling
+`SHAMap::walkSubTree`.  As `unshare()` walks the trie, nodes are not written to
+the database, and as `flushDirty` walks the trie nodes are written to the
+database. `walkSubTree` visits every node in the trie. This process must ensure
+that each node is only owned by this trie, and so performs a COW operation
+(`unshare()`) as it walks each node (from the root down).  This is done by the
+`preFlushNode` function.
+
+For each inner node encountered (starting with the root node), each of the
+children are inspected (from 1 to 16).  For each child, if it has a non-zero
+sequence number (unshareable), the child is first COW'd.  Then if the child is
+an inner node, we recurse down to that node's children.  Otherwise we've found a
+leaf node and that node is written to the database.  A count of each leaf node
+that is visited is kept.  The hash of the data in the leaf node is computed at
+this time, and the child is reassigned back into the parent inner node just in
+case the COW operation created a new pointer to this leaf node.
+
+After processing each node, the node is then marked as sharable again by setting
```

Comment:
Again, I think we need a section talking about the meaning of "shareable", as the concept is a bit unclear to me while reading this doc.